File: spike/tvos_playback/android/app/src/main/kotlin/co/podeo/opentv/tvos_playback/HostChannel.kt
package co.podeo.opentv.tvos_playback

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * The Android half of the host channel: a data directory and a keystore.
 *
 * Provider passwords are the reason this exists. Xtream puts the username and
 * password in the path of every stream URL, so the app has to keep the
 * password to play anything at all — which makes where it is kept a real
 * decision rather than a detail. It is not in the database: the schema holds
 * only a reference, and the secret lives here, encrypted with a key held in
 * the hardware-backed Android keystore and never leaving it.
 */
class HostChannel(private val context: Context) {

    fun attach(messenger: BinaryMessenger) {
        MethodChannel(messenger, "opentv/host").setMethodCallHandler(::handle)
    }

    /**
     * Built lazily and kept, because deriving the master key touches the
     * keystore and is slow enough to notice if it happened per call.
     */
    private val secrets: SharedPreferences by lazy {
        val key = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()

        EncryptedSharedPreferences.create(
            context,
            "opentv_credentials",
            key,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            // Internal storage: private to the app and durable until it is
            // uninstalled. Android has no equivalent of the tvOS purge, so
            // the catalogue can simply live here.
            "dataDirectory" -> result.success(context.filesDir.absolutePath)

            "writeSecret" -> {
                val reference = call.argument<String>("reference")
                val secret = call.argument<String>("secret")
                if (reference == null || secret == null) {
                    result.error("bad-args", "reference and secret required", null)
                    return
                }
                secrets.edit().putString(reference, secret).apply()
                result.success(null)
            }

            // A missing secret is an ordinary answer, not an error: clearing
            // the app's data leaves a source whose password is simply gone.
            "readSecret" -> result.success(
                call.argument<String>("reference")?.let { secrets.getString(it, null) }
            )

            "deleteSecret" -> {
                call.argument<String>("reference")?.let {
                    secrets.edit().remove(it).apply()
                }
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }
}

File: apps/opentv/apple/HostChannel.swift
import Flutter
import Foundation
import Security
import UIKit

/// The tvOS half of the host channel: a data directory and the Keychain.
///
/// Both answers differ from Android's in ways that matter to the app above.
final class HostChannel {

    /// The opentv:// link this launch was opened by, if any.
    ///
    /// Held rather than delivered as an event, for the reason the Android
    /// half holds one too: the link is what opened the app, so it has already
    /// happened before any Dart exists to hear about it. Cleared when read.
    static var pendingLink: String?

    /// The live channel, so a link arriving while the app is running can be
    /// pushed rather than waiting to be asked for.
    private static var channel: FlutterMethodChannel?

    /// Tells Dart about a link that arrived after the tree was already up.
    ///
    /// Pulling alone is not enough. `initialLink` is read once, when the tree
    /// comes up, which covers a cold launch and nothing else — a code scanned
    /// with the app already open set a pending link that nobody ever
    /// collected.
    static func announce(_ link: String) {
        channel?.invokeMethod("link", arguments: link)
    }

    static func attach(messenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(name: "opentv/host", binaryMessenger: messenger)
        channel.setMethodCallHandler { call, result in
            handle(call, result)
        }
        self.channel = channel
    }

    private static func handle(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {

        // UIKit's own idiom, shared verbatim with the iOS target.
        //
        // This file is compiled into both, which is the point: tvOS reports
        // itself to Dart as iOS, so nothing above this line can tell an Apple
        // TV from an iPhone. UIUserInterfaceIdiom can, and it is the same
        // answer on both platforms rather than two implementations that have
        // to be kept agreeing.
        case "initialLink":
            result(pendingLink)
            pendingLink = nil
            return

        case "deviceClass":
            switch UIDevice.current.userInterfaceIdiom {
            case .tv: result("television")
            case .pad: result("tablet")
            default: result("phone")
            }
            return

        // The caches directory, and not by preference.
        //
        // tvOS gives an app no directory that survives by right: there is a
        // small key-value store, and a cache the system may purge whenever it
        // wants space. A 284,000-row catalogue fits in neither, so on Apple TV
        // it is a cache by necessity and can vanish between launches. The Dart
        // side is written to expect that; the sync engine is resumable and
        // checkpointed per stage precisely so re-filling it is routine rather
        // than an incident.
        case "dataDirectory":
            let paths = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)
            guard let base = paths.first else {
                result(FlutterError(code: "no-directory",
                                    message: "tvOS returned no caches directory",
                                    details: nil))
                return
            }
            let directory = (base as NSString).appendingPathComponent("opentv")
            do {
                try FileManager.default.createDirectory(
                    atPath: directory, withIntermediateDirectories: true)
                result(directory)
            } catch {
                result(FlutterError(code: "no-directory",
                                    message: error.localizedDescription,
                                    details: nil))
            }

        case "writeSecret":
            guard let args = call.arguments as? [String: Any],
                  let reference = args["reference"] as? String,
                  let secret = args["secret"] as? String,
                  let data = secret.data(using: .utf8) else {
                result(FlutterError(code: "bad-args",
                                    message: "reference and secret required",
                                    details: nil))
                return
            }

            // Delete first rather than trying to update: SecItemAdd fails with
            // errSecDuplicateItem on a re-entered password, which is the most
            // ordinary case there is.
            SecItemDelete(query(for: reference) as CFDictionary)

            var attributes = query(for: reference)
            attributes[kSecValueData as String] = data
            // Never synchronised to other devices and never in a backup: this
            // is one television's copy of one provider's password.
            attributes[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly

            let status = SecItemAdd(attributes as CFDictionary, nil)
            guard status == errSecSuccess else {
                result(FlutterError(code: "keychain",
                                    message: "could not store the secret (OSStatus \(status))",
                                    details: nil))
                return
            }
            result(nil)

        case "readSecret":
            guard let args = call.arguments as? [String: Any],
                  let reference = args["reference"] as? String else {
                result(nil)
                return
            }
            var lookup = query(for: reference)
            lookup[kSecReturnData as String] = true
            lookup[kSecMatchLimit as String] = kSecMatchLimitOne

            var item: CFTypeRef?
            let status = SecItemCopyMatching(lookup as CFDictionary, &item)
            // A missing secret is an ordinary answer, not a failure: the
            // catalogue's directory can be purged, or the app's data cleared,
            // leaving a source whose password is simply gone.
            guard status == errSecSuccess, let data = item as? Data else {
                result(nil)
                return
            }
            result(String(data: data, encoding: .utf8))

        case "deleteSecret":
            if let args = call.arguments as? [String: Any],
               let reference = args["reference"] as? String {
                SecItemDelete(query(for: reference) as CFDictionary)
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private static func query(for reference: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "com.anthonyessaye.opentv.credentials",
            kSecAttrAccount as String: reference,
        ]
    }
}
